lib/vfscore/automount.c

```diff
@@ -48,36 +48,145 @@
 #include <uk/init.h>
 #include <uk/libparam.h>
 #include <uk/plat/memory.h>
+#include <sys/stat.h>
 
 #ifdef CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS
 #include <errno.h>
 #include <uk/config.h>
 #include <uk/arch/types.h>
-#include <sys/stat.h>
 #endif /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
 
+#if CONFIG_LIBVFSCORE_FSTAB
+#define LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP			':'
+#define LIBVFSCORE_FSTAB_UKOPTS_ARGS_SEP			','
+#endif /* CONFIG_LIBVFSCORE_FSTAB */
+
 struct vfscore_volume {
 	/* Volume source device */
 	const char *sdev;
 	/* Mount point absolute path */
-	const char *path;
+	char *path;
 	/* Corresponding filesystem driver name */
 	const char *drv;
 	/* Mount flags */
 	unsigned long flags;
 	/* Mount options */
 	const char *opts;
+#if CONFIG_LIBVFSCORE_FSTAB
+	/* Unikraft Mount options, see vfscore_mount_volume() */
+	char *ukopts;
+#endif /* CONFIG_LIBVFSCORE_FSTAB */
 };
 
+#if CONFIG_LIBVFSCORE_FSTAB
+/* Handle `mkmp` Unikraft Mount Option */
+static int vfscore_volume_process_ukopts_do_mkmp(char *path)
+{
+	char *pos, *prev_pos;
+	int rc;
+
+	UK_ASSERT(path);
+	UK_ASSERT(path[0] == '/');
+
+	pos = path;
+	do {
+		prev_pos = pos;
+		pos = strchr(pos + 1, '/');
+
+		if (pos) {
+			if (pos[0] == '\0')
+				break;
+
+			/* Zero out the next '/' */
+			*pos = '\0';
+		}
+
+		/* Do not allow `/./` or `/../` in the path. Also do not allow
+		 * overwriting .. or . files
+		 */
+		if (unlikely(prev_pos[1] == '.' &&
+			     /* /../ and /.. */
+			     ((prev_pos[2] == '.' &&
+			       (prev_pos[3] == '/' || prev_pos[3] == '\0')) ||
+				/* OR /./ and /. */
+			      (prev_pos[2] == '/' || prev_pos[2] == '\0')
+			     )))
+			uk_pr_err("Do not allow `/./` or `/../` in the path. "
+				  "Also do not allow overwriting .. or . "
+				  "files\n");
+
+		/* mkdir() with S_IRWXU */
+		rc = mkdir(path, 0700);
+		if (rc && errno != EEXIST)
+			return rc;
+
+		/* Restore current '/' */
+		if (pos)
+			*pos = '/';
+
+		/* Handle paths with multiple `/` */
+		while (pos && pos[1] == '/')
+			pos++;
+	} while (pos);
```

Comment:
It actually is for the last node of the path if it does not end with a `/`, which results in `strchr` returning a `NULL`.
E.g. say you have just entered the loop with the following:
`path = /path/to/file`, `pos = /file`, `prev_pos = /to/file`
In that case, `prev_pos` becomes `/file`, `strchr(pos + 1, '/');` returns `NULL` so `pos` becomes `NULL`. You now have: `path = /path/to/file`, `pos = NULL`, `prev_pos = /file`. I need the last `while` to still check for `pos != NULL` in case we have something like `/path/to/file///////` and then the last check on `pos != NULL` (the one of the `do { } while`) catches that case.

This is why in the previous version of this PR I had another `mkdir` outside the loop, to avoid having to do this step inside the loop.

Reply:
You are right. All good then.